src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -257,16 +288,18 @@ void addPartitionSensors(int partition) {
    */
   private class ProduceRunnable implements Runnable {
     private final int _partition;
+    private final String _key;
 
-    ProduceRunnable(int partition) {
+    ProduceRunnable(int partition, String key) {
       _partition = partition;
+      _key = key;
     }
 
     public void run() {
       try {
-        long nextIndex = _nextIndexPerPartition.get(_partition).get();
+        long nextIndex = _nextIndexPerKey.get(_key).get();
```

Comment:
It is a bit weird that `_nextIndexPerKey` is indexed by `key` while `_recordsProducedPerPartition` is still indexed by `partition`. In order to keep the change smaller and make the code style consistent, how about we still use `_nextIndexPerPartition`? We only need `key` to construct `BaseProducerRecord`.